### PR TITLE
feat: support UIActivityItemSource to customize share-sheet for iOS 13

### PR DESCRIPTION
**Pull request by miyabi**
# Overview

Share sheet was completely renewed in iOS13. It presents a preview of the link.
If this feature isn't supported, the appearance isn't nice.
So I implemented providing data to share by using UIActivityItemSource.
It allows also to share different data according to activities and to customize the share sheet.

# Test Plan

Create a new React Native project and change its App.js like example code.

# Example Code
```jsx:App.js
import React from 'react';
import {Platform, Text, TouchableWithoutFeedback, View} from 'react-native';
import Share from 'react-native-share';

const url = 'https://twitter.com/myb/status/1219906294976401408';
const title = 'Twitter';
const message = 'Here I am.';

const App: () => React$Node = () => {
  return (
    <View style={{flex: 1, justifyContent: 'center', alignItems: 'center'}}>
      <TouchableWithoutFeedback
        onPress={() => {
          const options = Platform.select({
            ios: {
              activityItemSources: [
                {
                  placeholderItem: { type: 'url', content: url },
                  item: {
                    default: { type: 'url', content: url },
                  },
                  subject: {
                    default: title,
                  },
                  linkMetadata: { originalUrl: url, url, title },
                },
                {
                  placeholderItem: { type: 'text', content: message },
                  item: {
                    default: { type: 'text', content: message },
                    message: null,
                  },
                },
              ],
            },
            default: {
              title,
              subject: title,
              message: `${message} ${url}`,
            },
          });
          Share.open(options).catch(e => console.log(e));
        }}>
        <Text style={{ fontSize: 16, fontWeight: 'bold' }}>
          Share with activity item sources
        </Text>
      </TouchableWithoutFeedback>
    </View>
  );
}

export default App;
```

# Demo
RPReplay_Final1580215842 2020-01-28 9_54_46 PM

### CODE REVIEW

**Comment by MateusAndrade** [comment]
Hi @miyabi , you can update your base with the master version?

**Comment by miyabi** [comment]
@MateusAndrade I did.

**Comment by MateusAndrade** [comment]
Hey, @miyabi sorry for the late reply. When I was trying to run your changes at the example App, i got the following error:

```
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_RNShareActivityItemSource", referenced from:
      objc-class-ref in libRNShare.a(RNShare.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

File: ios/RNShareActivityItemSource.m [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        if (videoURL) {
+            linkMetadata.videoProvider = [[NSItemProvider alloc] initWithContentsOfURL:videoURL];
+        }
+        return linkMetadata;
+    }
+    return nil;
+}
+
++ (nullable NSString *)keyForActivityType:(UIActivityType)activityType {
+    if ([activityType isEqual:UIActivityTypeAddToReadingList]) {
```
Comment (MateusAndrade) [inline ios/RNShareActivityItemSource.m]: What do you think on using something like:

```
switch (activityType)
   {
      case isEqual:UIActivityTypeAddToReadingList:
        return @"addToReadingList";
        break;
}
```

**Comment by MateusAndrade** [comment]
Thanks!

File: ios/RNShareActivityItemSource.m (hunk 1 above)
Comment (miyabi) [reply ios/RNShareActivityItemSource.m]: @MateusAndrade Sorry for my late reply. `switch` statement doesn't work with string in Objective-C. There are some ways to do equivalent, for example, using macros or using dictionary of blocks, but I think they are not cool. Anyway, thank you for merge!

**Comment by evancloutier** [comment]
@MateusAndrade 👋 Are there plans to release this feature soon?

**Comment by MateusAndrade** [comment]
> @MateusAndrade 👋 Are there plans to release this feature soon?

Sure! I was planning to release that on Monday. But I don't get time to test the master version on both android and iOS. If you can provide any useful info about our `master` stability, this would help me a lot.  ❤️ ❤️ ❤️

**Comment by evancloutier** [comment]
@MateusAndrade I'll try and get around to both before Monday 👍

**Comment by MateusAndrade** [comment]
Hey @evancloutier i just published a pre-release for that: https://github.com/react-native-community/react-native-share/releases/tag/v3.1.0. 💃 💃 💃

**Comment by evancloutier** [comment]
@MateusAndrade iOS is looking good in terms of functionality. We're missing type definitions for `activityItemSources` however. Did we want to add those before releasing?

**Comment by MateusAndrade** [comment]
> @MateusAndrade iOS is looking good in terms of functionality. We're missing type definitions for `activityItemSources` however. Did we want to add those before releasing?

We need to add that to https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react-native-share. I can look on that if you send a PR adding this types. 😄

**Comment by evancloutier** [comment]
@miyabi Is there documentation I can refer to for the different types of UIActivityItemSources that are available? For placeholders, etc. I'm not too familiar with Apple docs so I don't want to miss anything 😅

**Comment by miyabi** [comment]
@evancloutier Actually, I've added type annotations for flow type. Is it helpful?
https://github.com/react-native-community/react-native-share/pull/693/files#diff-168726dbe96b3ce427e7fedce31bb0bcR110

I referred these Apple docs.

UIActivityItemSource (each methods correspond to items like: activityViewControllerPlaceholderItem -> placeholderItem):
https://developer.apple.com/documentation/uikit/uiactivityitemsource

LinkMetadata:
https://developer.apple.com/documentation/linkpresentation/lplinkmetadata

ActivityType:
https://developer.apple.com/documentation/uikit/uiactivity/activitytype

**Comment by evancloutier** [comment]
Yep! I'll push up a PR tomorrow morning @MateusAndrade. Thanks @miyabi.

**Pull request merged**